### PR TITLE
407 feature request fcrackzip tools

### PR DESCRIPTION
**Pull request by namanbkshi**
This Pull request is generated with the intentions of solving two issues namely: 
1. 407 feature request fcrackzip tools #634 and 
2. [Bug Fix (Baseline)]: Fcrackzip #586

For the feature request:

1.Set 'Show cracked password' to be enabled by default: Initialized the checkedUnzip state with true, so the checkbox for displaying the cracked password is enabled by default.

2. Ensure Charset List remains active: Removed logic that disabled Charset List options based on other checkbox states. Now, all Charset List options remain active and selectable regardless of other option states, ensuring consistent usability.

3.Adjusted the output behavior to always display the cracked password, especially when using the Dictionary Attack Method.
   - Implemented parsing of the command output to extract the cracked password using a regular expression.
   - If a cracked password is found, it is now displayed in the output along with the rest of the output.
   - Additionally, if the "Show cracked password" option is selected, the cracked password is saved to disk.

Improved the baseline by improving the comments & docstrings along with implementing the following changes:

Certainly! Here's a breakdown of the changes in bullet points:

- **Imports Refactoring**:
  - Updated import statements to reflect changes in dependencies.

- **Component Structure**:
  - Retains the same state variables and form handling logic.

- **Dependency Check**:
  - Introduces a new state variable `dependencyAvailable` to track dependency availability.
  - Asynchronously checks for dependency availability on component mount.

- **Dependency Availability UI**:
  - Displays a message indicating dependency unavailability if `fcrackzip` is not found.

- **UI Improvements**:
  - Loading overlay and cancel button are now rendered using a separate `LoadingOverlayAndCancelButton` component.
  - User guide rendering logic moved to a separate `RenderComponent` component, which receives props for title, description, steps, tutorial link, and source code link.

- **Clear Output**:
  - Retains the function to clear output for users.

### CODE REVIEW

**Review by bradleybogg — CHANGES_REQUESTED** [review]
@namanbkshi 

Please complete a PR that reflects what has been changed.
https://github.blog/2015-01-21-how-to-write-the-perfect-pull-request/

**Comment by bradleybogg** [comment]
No update on previous request.

**Comment by namanbkshi** [comment]
> No update on previous request.

Hello Bradley ! This pull request was made for the implementation of feature request but in the weekly meeting I was told to implement the baseline as well in the same request. 

So perhaps you want me to create another pull request ?

**Comment by namanbkshi** [comment]
Hi. I am done with the changes. Thanks in advance.

File: src/components/Fcrackzip/Fcrackzip.tsx [hunk 1]
```diff
@@ -1,19 +1,29 @@
-import { Button, LoadingOverlay, Stack, TextInput, NativeSelect, Checkbox } from "@mantine/core";
+import { Button, Stack, TextInput, NativeSelect, Checkbox } from "@mantine/core";
 import { useForm } from "@mantine/form";
 import { useEffect, useCallback, useState } from "react";
 import { CommandHelper } from "../../utils/CommandHelper";
 import ConsoleWrapper from "../ConsoleWrapper/ConsoleWrapper";
-import { UserGuide } from "../UserGuide/UserGuide";
+import { RenderComponent } from "../UserGuide/UserGuide";
 import { SaveOutputToTextFile } from "../SaveOutputToFile/SaveOutputToTextFile";
```
Comment (michaeljpigott) [inline src/components/Fcrackzip/Fcrackzip.tsx:7]: As part of baselining, this should be changed to aveOutputToTextFile_v2. 
```suggestion
import { SaveOutputToTextFile_v2 } from "../SaveOutputToFile/SaveOutputToTextFile";
```

File: src/components/Fcrackzip/Fcrackzip.tsx [hunk 2]
```diff
@@ -1,19 +1,29 @@
...
+import { checkAllCommandsAvailability } from "../../utils/CommandAvailability";
+import { LoadingOverlayAndCancelButton } from "../OverlayAndCancelButton/OverlayAndCancelButton";
```
Comment (michaeljpigott) [inline src/components/Fcrackzip/Fcrackzip.tsx:10]: You are missing LoadingOverlayAndCancelButton.
```suggestion
import { LoadingOverlayAndCancelButton } from "../OverlayAndCancelButton/OverlayAndCancelButton";
```

File: src/components/Fcrackzip/Fcrackzip.tsx [hunk 3]
```diff
@@ -1,19 +1,29 @@
...
 
-const title = "Fcrackzip Tool";
-const description_guide =
-    "Fcrackzip is a tool for cracking password of a protected zip file.\n\n" +
-    "Step 1: Input path of the zip file.\n\n" +
-    "Step 2: Select an Attack Mode (Dictionary or Brute Force)\n\n" +
-    "Step 3: You can save output by checking 'Save Output to File'\n\n" +
-    "Step 4: Click START CRACKING!\n\n";
+// Component title
```
Comment (michaeljpigott) [inline src/components/Fcrackzip/Fcrackzip.tsx:11]: These component constants need to be moved to within the Fcrackzip function.

File: src/components/Fcrackzip/Fcrackzip.tsx [hunk 4]
```diff
@@ -1,19 +1,29 @@
...
+const title = "Fcrackzip";
+// Description of the component
+const description_guide = "Fcrackzip is a tool for cracking password of a protected zip file.";
```
Comment (michaeljpigott) [inline src/components/Fcrackzip/Fcrackzip.tsx:14]: Add the
```suggestion
const description_guide = "Fcrackzip is a tool for cracking the password of a protected zip file.";
```

File: src/components/Fcrackzip/Fcrackzip.tsx [hunk 5]
```diff
@@ -1,19 +1,29 @@
...
+// Additional props for the RenderComponent
+const steps =
+    "Step 1: Input path of the zip file.\n" +
+    "Step 2: Select an Attack Mode (Dictionary or Brute Force)\n" +
+    "Step 3: You can save output by checking 'Save Output to File'\n" +
+    "Step 4: Click START CRACKING!";
```
Comment (michaeljpigott) [inline src/components/Fcrackzip/Fcrackzip.tsx:20]: Make this consistent with the actual button, which is Start Cracking!
```suggestion
    "Step 4: Click Start Cracking!";
```

File: src/components/Fcrackzip/Fcrackzip.tsx [hunk 6]
```diff
@@ -49,31 +69,32 @@ const Fcrackzip = () => {
             charSet: "",
         },
     });
-    // Ask per request of the tester, the checkzip will not disable the Setchar anymore but the logic will only be commented out
-    //if needed in the future we can uncomment the change back
-    // the useeffect is not really useful right now and maybe remove in future implementation
+
+    // Check for dependency availability when the component mounts
```
Comment (michaeljpigott) [inline src/components/Fcrackzip/Fcrackzip.tsx:73]: Full stop. 
```suggestion
    // Check for dependency availability when the component mounts.
```

File: src/components/Fcrackzip/Fcrackzip.tsx [hunk 7]
```diff
@@ -49,31 +69,32 @@ const Fcrackzip = () => {
...
 
-        //if (checkedUnzip) {
-        //setCharSetDisabled(true);
-        //} else {
-        //setCharSetDisabled(false);
-        //}
-    }, [useCharsetLowercase, useCharsetUppercase, useCharsetNumeric, checkedUnzip]);
+    // Function to check dependency availability
+    const checkDependencyAvailability = async () => {
+        const isAvailable = await checkAllCommandsAvailability(dependencies);
```
Comment (michaeljpigott) [inline src/components/Fcrackzip/Fcrackzip.tsx:80]: This doesn't appear to be working in the app. Please test and review airbase-ng to see how the code should be implemented. 

To test, uninstall fcrackzip on your local machine, open DDT and navigate to the fcrackzip tool. If it is working, then the installation modal should open and ask you to install the dependency. 

It did not work for me after uninstalling fcrackzip.

File: src/components/Fcrackzip/Fcrackzip.tsx [hunk 8]
```diff
@@ -135,118 +149,134 @@ const Fcrackzip = () => {
...
+            steps={steps}
+            tutorial={tutorial}
+            sourceLink={sourceLink}
+        >
+            {!dependencyAvailable && (
                 <div>
-                    <Button variant="outline" color="red" style={{ zIndex: 1001 }} onClick={handleCancel}>
-                        Cancel
-                    </Button>
+                    <p>Dependency not available. Please install it.</p>
```
Comment (michaeljpigott) [inline src/components/Fcrackzip/Fcrackzip.tsx:191]: You should not need this once you get checkAllCommandsAvailability working correctly.

File: src/components/Fcrackzip/Fcrackzip.tsx [hunk 9]
```diff
@@ -135,118 +149,134 @@ const Fcrackzip = () => {
...
-                {SaveOutputToTextFile(output)}
-                <Button type={"submit"}>Start Cracking!</Button>
-
-                <ConsoleWrapper output={output} clearOutputCallback={clearOutput} />
-                <Button onClick={clearOutput}>Clear Output</Button>
-            </Stack>
-        </form>
+                        </>
+                    )}
+                    {SaveOutputToTextFile(output)}
```
Comment (michaeljpigott) [inline src/components/Fcrackzip/Fcrackzip.tsx:273]: This should be SaveOutputToTextFile_v2
```suggestion
                  {SaveOutputToTextFile_v2(output, allowSave, hasSaved, handleSaveComplete)}
```

File: src/components/Fcrackzip/Fcrackzip.tsx [hunk 10]
```diff
@@ -135,118 +149,134 @@ const Fcrackzip = () => {
...
+                    <Button type={"submit"}>Start Cracking!</Button>
+                    <ConsoleWrapper output={output} clearOutputCallback={clearOutput} />
+                    <Button onClick={clearOutput}>Clear Output</Button>
```
Comment (michaeljpigott) [inline src/components/Fcrackzip/Fcrackzip.tsx:276]: Remove not required. 
image

**Review by michaeljpigott — CHANGES_REQUESTED** [review]
Hi @namanbkshi 

There is a bit to work on here, including:
1. Ensuring you use the SaveOutputToTextFile_v2 component rather than SaveOutputToTextFile
2. Import the LoadingOverlayAndCancelButton component
3. Move the component constants to within the fcrackzip function 
4. Fix up the dependency checker

Kind regards 

Michael

**Comment by namanbkshi** [comment]
Hey Michael, I have implemented the changes the new commit. Thanks for your help.

**Review by michaeljpigott — APPROVED** [review]

**Pull request merged**